cou/steps/plan.py

```diff
@@ -375,6 +379,65 @@ async def generate_plan(analysis_result: Analysis, args: CLIargs) -> UpgradePlan
     return plan
 
 
+async def filter_hypervisors_machines(args: CLIargs, analysis_result: Analysis) -> set[Machine]:
+    """Filter the hypervisors to generate plan and upgrade.
+
+    :param args: CLI arguments
+    :type args: CLIargs
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :return: hypervisors filtered to generate plan and upgrade.
+    :rtype: set[Machine]
+    """
+    hypervisors_machines = await _get_hypervisors_machines_possible_to_upgrade(
+        args.force, analysis_result
+    )
+
+    if cli_machines := args.machines:
+        return {machine for machine in hypervisors_machines if machine.machine_id in cli_machines}
+
+    if cli_hostnames := args.hostnames:
+        return {machine for machine in hypervisors_machines if machine.hostname in cli_hostnames}
+
+    if cli_azs := args.availability_zones:
+        return {machine for machine in hypervisors_machines if machine.az in cli_azs}
+
+    return hypervisors_machines
+
+
+async def _get_hypervisors_machines_possible_to_upgrade(
+    cli_force: bool, analysis_result: Analysis
+) -> set[Machine]:
+    """Get the hypervisors that are possible to upgrade.
+
+    :param cli_force: If force is used, it gets all hypervisors, otherwise just the empty ones
+    :type cli_force: bool
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :return: Set of nova-compute units to upgrade
+    :rtype: set[Machine]
+    """
+    hypervisors_units = _get_hypervisors_units(analysis_result.apps_data_plane)
+
+    if cli_force:
+        return {unit.machine for unit in hypervisors_units}
+
+    hypervisors_units_name = [unit.name for unit in hypervisors_units]
```

Comment:
nit: any reason we are using list instead of set here? We are using set everywhere else in this function so was wondering if I missed anything.

Reply:
When I use the asyncio.gather the input needs to match with the exactly order of the inputs. Set are a little bit unpredictable the sequence, so using list I feel more safe from bugs.

What I'm trying to say that in a input like ["A", "B", "C"] in a function `foo(input)` I would like to be sure that the result of `foo(A)` won't be considered as output of `foo(B)` or `foo(C)`. Playing around with set it seems that they always return the same sequence when you interact, so you might be right that its not necessary.

Reply:
Right, since we need to `zip` units and instance count together. Makes sense :ok_hand: